hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/CustomFeeAssessmentStep.java

```diff
@@ -207,9 +199,36 @@ public CustomFeeAssessmentResult assessFees(
         if (!hbarTransfers.isEmpty() || !tokenTransfers.isEmpty()) {
             assessedTxns.add(txnToAssess);
         }
+        // If custom fee adjustments were incurred, we need to validate the total number
+        // of balance adjustments and ownership changes is within our limits
+        if (!customFeesAssessed.isEmpty()) {
+            final var maxBalanceChanges =
+                    config.getConfigData(LedgerConfig.class).xferBalanceChangesMaxLen();
+            validateTrue(
+                    numUniqueAdjustmentsIn(assessedTxns) <= maxBalanceChanges,
+                    CUSTOM_FEE_CHARGING_EXCEEDED_MAX_ACCOUNT_AMOUNTS);
+        }
         return new CustomFeeAssessmentResult(assessedTxns, customFeesAssessed);
     }
 
+    private int numUniqueAdjustmentsIn(final List<CryptoTransferTransactionBody> assessedTxns) {
+        var numOwnershipChanges = 0;
+        final Set<AccountID> uniqueHbarAdjustments = new HashSet<>();
+        final Set<Pair<AccountID, TokenID>> uniqueTokenAdjustments = new HashSet<>();
+        for (final var txn : assessedTxns) {
+            for (final var aa : txn.transfersOrElse(TransferList.DEFAULT).accountAmountsOrElse(emptyList())) {
+                uniqueHbarAdjustments.add(aa.accountID());
+            }
+            for (final var xfer : txn.tokenTransfersOrElse(emptyList())) {
+                for (final var aa : xfer.transfersOrElse(emptyList())) {
+                    uniqueTokenAdjustments.add(Pair.of(aa.accountID(), xfer.token()));
+                }
+                numOwnershipChanges += xfer.nftTransfersOrElse(emptyList()).size();
+            }
```

Comment:
Nit: It could it be easier on the eyes to refactor these for loops into 2 separate methods something along the lines of countUniqueHbarAdjustments( ) and countUniqueTokenAdjustments( ) ? : )

Reply:
Ah good idea, will change this in upcoming PR.